File: content/examples/other/leetcode/problems/reorder-list/example-1/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
  static class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
      this.val = val;
    }

    ListNode(int val, ListNode next) {
      this.val = val;
      this.next = next;
    }
  }

  static void reorderListWithArray(ListNode head) {
    List<ListNode> nodes = new ArrayList<>();
    for (ListNode cur = head; cur != null; cur = cur.next) nodes.add(cur);
    int left = 0;
    int right = nodes.size() - 1;
    while (left < right) {
      nodes.get(left).next = nodes.get(right);
      left += 1;
      if (left == right) break;
      nodes.get(right).next = nodes.get(left);
      right -= 1;
    }
    if (!nodes.isEmpty()) nodes.get(left).next = null;
  }

  static List<Integer> toArray(ListNode head) {
    List<Integer> out = new ArrayList<>();
    for (ListNode cur = head; cur != null; cur = cur.next) out.add(cur.val);
    return out;
  }

  public static void main(String[] args) {
    ListNode a = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
    reorderListWithArray(a);
    System.out.println(toArray(a));
  }
}

File: content/examples/other/leetcode/problems/reorder-list/example-1/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reorder_list_with_array(head):
    nodes = []
    cur = head
    while cur:
        nodes.append(cur)
        cur = cur.next
    left, right = 0, len(nodes) - 1
    while left < right:
        nodes[left].next = nodes[right]
        left += 1
        if left == right:
            break
        nodes[right].next = nodes[left]
        right -= 1
    if nodes:
        nodes[left].next = None
    return head


def to_array(head):
    out = []
    cur = head
    while cur:
        out.append(cur.val)
        cur = cur.next
    return out


if __name__ == "__main__":
    a = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
    reorder_list_with_array(a)
    print(to_array(a))
